android/modules/database/src/java/ti/modules/titanium/database/TiDatabaseProxy.java

```diff
@@ -18,145 +19,244 @@
+					if (cursor.getColumnCount() > 0) {
+						result = new TiResultSetProxy(cursor);
+						if (result.isValidRow()) {
+							result.next();
 						}
-					} else {
-						c.close();
-						c = null;
-						rs = null;
 					}
-				} else {
-					// Leaving for historical purposes, but walking through several different
-					// types of statements never hit this branch. (create, drop, select, pragma)
-					rs = new TiResultSetProxy(null); // because iPhone does it this way.
 				}
+
+				// Query does not return result, use execSQL()
 			} else {
-				Object[] newArgs = null;
-				if (sqlArgs != null) {
-					newArgs = new Object[sqlArgs.length];
-					for (int i = 0; i < sqlArgs.length; i++) {
-						if (sqlArgs[i] instanceof TiBlob) {
-							newArgs[i] = ((TiBlob) sqlArgs[i]).getBytes();
-						} else {
-							newArgs[i] = TiConvert.toString(sqlArgs[i]);
-						}
-					}
-				}
-				db.execSQL(sql, newArgs);
+				db.execSQL(query, parameterObjects);
 			}
-		} catch (SQLException e) {
-			String msg = "Error executing sql: " + e.getMessage();
-			Log.e(TAG, msg, e);
-			if (c != null) {
+		} finally {
+
+			// Cleanup result.
+			if (cursor != null) {
 				try {
-					c.close();
-				} catch (SQLException e2) {
-					// Ignore
+					cursor.close();
+				} catch (Exception e) {
+					// Ignore...
 				}
 			}
-			throw e;
 		}
 
-		return rs;
+		return result;
+	}
+
+	/**
+	 * Asynchronously execute a single SQL query.
+	 * @param callback Result callback for query execution.
+	 * @param query SQL query to execute on database.
+	 * @param parameters Parameters for `query`
+	 */
+	@Kroll.method
+	public void executeAsync(final KrollFunction callback, final String query, final Object... parameters)
+	{
+		executingQueue.set(true);
+		try {
+			queue.put(new Runnable() {
+				@Override
+				public void run()
+				{
+					final TiResultSetProxy result = execute(query, parameters);
+					callback.callAsync(getKrollObject(), new Object[] { result });
+				}
+			});
+		} catch (InterruptedException e) {
+		}
+	}
+
+	/**
+	 * Synchronously execute a multiple SQL queries.
+	 * @param queries SQL queries to execute on database.
+	 */
+	@Kroll.method
+	public Object[] executeAll(final String[] queries)
+	{
+		ArrayList<TiResultSetProxy> results = new ArrayList<>();
+		for (String query : queries) {
+			final TiResultSetProxy result = execute(query);
+			results.add(result);
+		}
+		return results.toArray();
+	}
+
+	/**
+	 * Asynchronously execute a multiple SQL queries.
+	 * @param callback Result callback for query execution.
+	 * @param queries SQL queries to execute on database.
+	 */
+	@Kroll.method
+	public void executeAllAsync(final KrollFunction callback, final String[] queries)
+	{
+		executingQueue.set(true);
+		try {
+			queue.put(new Runnable() {
+				@Override
+				public void run()
+				{
+					final Object[] results = executeAll(queries);
+					callback.callAsync(getKrollObject(), new Object[] { results });
```

Comment:
does results need to be "double-wrapped" here? I assume otherwise it'd expand out and the callback would get separate arguments for each results rather than a single array?

Reply:
Yes it does, each element of the `new Object[] { }` is a parameter in the callback. This will return a callback with a parameter containing an array of results.